### ISSUE TITLE
Please implement batching rule for torch.nn.functional.multi_margin_loss

### ISSUE THREAD

**Issue by tiendungchs** [pytorch/pytorch#139241]
### 🚀 The feature, motivation and pitch

I'm working on a metric to measure the efficiency of a neural network. My approach requires computing per-example gradient using multi_margin_loss.
Like evaluating the test accuracy of a model, the more examples used, the better the estimation. I would like to measure the efficiency of a lot of models using their entire validation set, so a speed boost would save so much time.

### Alternatives

In fact, my original idea is really simple, the loss is simply the output with the additive inverse of argmax value:
```
  m = torch.ones_like(output)
  m[label] = -1
  loss = output * m + 2
```
I've tried every way I know (indexing, scatter, one_hot, arange) to do this: `m[label] = -1`. However, I always run into problems with indexing BatchedTensor.

If vmap apply a function over a batch of data, I think a per-example function shouldn't have to deal with BatchedTensor in the first place. I'm new to all of this so I don't know if I made a mistake, or what is needed in this case, maybe the ability to index, or batching rule for scatter? 

I don't know which is the solution, but I feel like solving this gives more flexibility than implementing batching rule for any other function.

**Comment by Chillee** [comment]
@tiendungchs Generally speaking, there are some specific rules when it comes to vmapping with scatter ops. Do you have some example code and perhaps I can help?
